### DISCUSSION TITLE
Migration from Bluepy

### DISCUSSION THREAD

**Discussion by iPsych** [hbldh/bleak#1235] [Q&A]
Dear all.
I am trying to implement bleak to code built with bluepy.
Among bluepy functions below, is there any good migration example?
-Scanner
-DefaultDelegate
-AssignedNumbers

**Comment by StefJar** [comment]
I went through this as well
Imho just switch fully to bleak because it works directly on dbus and not via a c-cmdline program. It is faster and you don't need to adjust permissions like this https://stackoverflow.com/questions/59786226/how-to-setup-the-enviroment-that-bluepy-can-scan-without-sudo

Only thing I saw is that bleak does not implement any vendor specific gatt services. Means if bluez does not know that service uuid it won't create the service & character objects. Its currently in the discussion - feel free to add your experience https://github.com/hbldh/bleak/discussions/1204

The examples are in the "example" subdir. A good one is https://github.com/hbldh/bleak/blob/develop/examples/two_devices.py

Some words on your problems:
Scanner:
```
async def _scann(scannT_sec: float = 3.0):
    devices = await bleak.BleakScanner.discover(timeout=scannT_sec, return_adv=True)
    devL = []

    for addr, (bleDev, advData) in devices.items():
        devName = advData.local_name
        if (devName != None) and (devName == "myDevice"):
            devL.append(addr)
    return devL
```

DefaultDelegate & AssignedNumbers: You don't need this. Bluez is doing that in background and exposing the results as dbus objects. You might want to check `bluetoothctl`. Everything you see there is 1:1 mirrored of the bluez dbus stuff.
